File: app/src/main/java/com/gigabytedevelopersinc/app/CometOTP/Activities/AboutActivity.java
/*
 * Copyright (C) 2017-2018 Jakob Nixdorf
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gigabytedevelopersinc.app.CometOTP.Activities;

import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.ColorFilter;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.CardView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.ViewStub;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.github.aakira.expandablelayout.ExpandableLayoutListenerAdapter;
import com.github.aakira.expandablelayout.ExpandableLinearLayout;

import com.gigabytedevelopersinc.app.CometOTP.R;
import com.gigabytedevelopersinc.app.CometOTP.Utilities.Constants;
import com.gigabytedevelopersinc.app.CometOTP.Utilities.Tools;

public class AboutActivity extends BaseActivity {
    private static final String GITHUB_URI = "https://github.com/gigabytedevelopers/CometOTP";
    private static final String WHATSAPP_URI = "https://chat.whatsapp.com/HvUjQXAkMne3hjSiy1sUpc";
    private static final String CHANGELOG_URI = GITHUB_URI + "/blob/master/CHANGELOG.md";
    private static final String MIT_URI = GITHUB_URI + "/blob/master/LICENSE.txt";

    private static final String AUTHOR1_GITHUB = "https://github.com/gigabytedevelopers";
    private static final String AUTHOR1_PAYPAL = "https://paypal.me/gigabtedevelopers";

    /*private static final String AUTHOR2_GITHUB = "https://github.com";*/
    private static final String AUTHOR2_APP = "https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2";

    private static final String CONTRIBUTORS_URI = GITHUB_URI + "/graphs/contributors";
    private static final String TRANSLATORS_URI = GITHUB_URI + "/blob/master/README.md#translators";

    private static final String BUGREPORT_URI = GITHUB_URI + "/issues";
    private static final String TRANSLATE_URI = "http://gigabytedevelopers.oneskyapp.com/collaboration/project/296138";

    static final int[] imageResources = {
            R.id.aboutImgVersion, R.id.aboutImgLicense, R.id.aboutImgChangelog, R.id.aboutImgSource,
            R.id.aboutImgOpenSource, R.id.aboutImgAuthor1, R.id.aboutImgAuthor2, R.id.aboutImgContributors,
            R.id.aboutImgTranslators, R.id.aboutImgBugs, R.id.aboutImgTranslate
    };

    static long lastTap = 0;
    static int taps = 0;
    static Toast currentToast = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTitle(R.string.about_activity_title);
        setContentView(R.layout.activity_container);

        Toolbar toolbar = findViewById(R.id.container_toolbar);
        setSupportActionBar(toolbar);

        ViewStub stub = findViewById(R.id.container_stub);
        stub.setLayoutResource(R.layout.content_about);
        View v = stub.inflate();

        ColorFilter filter = Tools.getThemeColorFilter(this, android.R.attr.textColorSecondary);
        for (int i : imageResources) {
            ImageView imgView = v.findViewById(i);
            imgView.getDrawable().setColorFilter(filter);
        }

        String versionName = "";
        try {
            PackageInfo packageInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            versionName = packageInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }

        LinearLayout versionLayout = v.findViewById(R.id.about_layout_version);

        versionLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                long thisTap = System.currentTimeMillis();

                if (thisTap - lastTap < 500) {
                    taps = taps + 1;

                    if (currentToast != null && taps <= 7)
                        currentToast.cancel();

                    if (taps >= 3 && taps <= 7)
                        currentToast = Toast.makeText(getBaseContext(), String.valueOf(taps), Toast.LENGTH_SHORT);

                    if (taps == 7) {
                        if (settings.getSpecialFeatures())
                            Snackbar.make(findViewById(R.id.about), R.string.about_toast_special_features_enabled, Snackbar.LENGTH_LONG).show();
                            //currentToast = Toast.makeText(getBaseContext(), R.string.about_toast_special_features_enabled, Toast.LENGTH_LONG);
                        else
                            enableSpecialFeatures();
                    }

                    if (currentToast != null)
                        currentToast.show();
                } else {
                    taps = 0;
                }

                lastTap = thisTap;
            }
        });

        TextView version = v.findViewById(R.id.about_text_version);
        version.setText(versionName);

        LinearLayout license = v.findViewById(R.id.about_layout_license);
        LinearLayout changelog = v.findViewById(R.id.about_layout_changelog);
        LinearLayout source = v.findViewById(R.id.about_layout_source);
        LinearLayout licenses = v.findViewById(R.id.about_layout_licenses);
        license.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(MIT_URI);
            }
        });
        changelog.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final AlertDialog.Builder builder = new AlertDialog.Builder(AboutActivity.this);
                builder.setTitle("ChangeLog");
                builder.setMessage(R.string.chagelog);
                builder.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                    }
                });
                builder.create().show();
            }
        });
        source.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(WHATSAPP_URI);
            }
        });
        licenses.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                showLicenses();
            }
        });

        TextView author1GitHub = v.findViewById(R.id.about_author1_github);
        TextView author1Paypal = v.findViewById(R.id.about_author1_paypal);

        author1GitHub.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    openURI(AUTHOR1_GITHUB);
                } catch(Exception ignored) {
                    copyToClipboard(AUTHOR1_GITHUB);
                }
            }
        });
        author1Paypal.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    openURI(AUTHOR1_PAYPAL);
                } catch(Exception ignored) {
                    copyToClipboard(AUTHOR1_PAYPAL);
                }
            }
        });

        /*TextView author2GitHub = v.findViewById(R.id.about_author2_github);*/
        TextView author2App = v.findViewById(R.id.about_author2_app);
        /*author2GitHub.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(AUTHOR2_GITHUB);
            }
        });*/
        author2App.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(AUTHOR2_APP);
            }
        });

        LinearLayout contributors = v.findViewById(R.id.about_layout_contributors);
        LinearLayout translators = v.findViewById(R.id.about_layout_translators);
        contributors.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(CONTRIBUTORS_URI);
            }
        });
        translators.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(TRANSLATORS_URI);
            }
        });

        LinearLayout bugReport = v.findViewById(R.id.about_layout_bugs);
        LinearLayout translate = v.findViewById(R.id.about_layout_translate);
        bugReport.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final Intent feedback = new Intent(Intent.ACTION_SEND);
                feedback.setType("text/html");
                feedback.putExtra(Intent.EXTRA_EMAIL, new String[]{ getString(R.string.feedback_email_address)});
                feedback.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback_email_subject));
                feedback.putExtra(Intent.EXTRA_TEXT, getString(R.string.feedback_email_message));
                startActivity(Intent.createChooser(feedback, getString(R.string.feedback_email_title)));
            }
        });
        translate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                openURI(TRANSLATE_URI);
            }
        });

        final Button expandButton = v.findViewById(R.id.thumb_expand_button);
        final CardView expand = v.findViewById(R.id.thumb_expand);
        final ExpandableLinearLayout expandLayout = v.findViewById(R.id.thumb_disclaimer);

        expand.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                expandLayout.toggle();
            }
        });

        expandButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                expandLayout.toggle();
            }
        });

        expandLayout.setListener(new ExpandableLayoutListenerAdapter() {
            @Override
            public void onOpened() {
                super.onOpened();
                expandButton.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, R.drawable.ic_arrow_up, 0);
            }

            @Override
            public void onClosed() {
                super.onClosed();
                expandButton.setCompoundDrawablesRelativeWithIntrinsicBounds(0, 0, R.drawable.ic_arrow_down, 0);
            }
        });
    }

    private void enableSpecialFeatures() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.about_title_special_features)
                .setMessage(R.string.about_dialog_special_features)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        settings.setSpecialFeatures(true);

                        Snackbar.make(findViewById(R.id.about), R.string.about_toast_special_features, Snackbar.LENGTH_LONG).show();
                        //Toast.makeText(getBaseContext(), R.string.about_toast_special_features, Toast.LENGTH_LONG).show();
                    }
                })
                .setNegativeButton(android.R.string.cancel, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {}
                })
                .create()
                .show();
    }

    // Go back to the main activity
    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }

    @Override
    public void onBackPressed() {
        finish();
        super.onBackPressed();
    }

    public void openURI(String uri) {
        Intent openURI = new Intent(Intent.ACTION_VIEW);
        openURI.setData(Uri.parse(uri));
        startActivity(openURI);
    }
    
    public void copyToClipboard(String uri) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("CometOTP", uri);
        clipboard.setPrimaryClip(clip);
        Snackbar.make(findViewById(R.id.about), getString(R.string.about_toast_copied_to_clipboard), Snackbar.LENGTH_LONG).show();
        //Toast.makeText(this, getString(R.string.about_toast_copied_to_clipboard), Toast.LENGTH_SHORT).show();
    }


    public void showLicenses() {
        Intent licensesIntent = new Intent(this, LicensesActivity.class);
        startActivityForResult(licensesIntent, Constants.INTENT_MAIN_LICENSES);
        /*String backgroundColor = Tools.getCSSRGBAString(Tools.getThemeColor(this, R.attr.colorBackgroundFloating));
        String textColor = Tools.getCSSRGBAString(Tools.getThemeColor(this, android.R.attr.textColorPrimary));
        String textColorSecondary = Tools.getCSSRGBAString(Tools.getThemeColor(this, android.R.attr.textColorSecondary));
        String cssFormat = getString(R.string.custom_notices_style, backgroundColor, textColor, textColorSecondary);
        LicensesDialog dialog = new LicensesDialog.Builder(this)
                .setNotices(R.raw.licenses)
                .setTitle(R.string.about_label_licenses)
                .setShowFullLicenseText(false)
                .setIncludeOwnLicense(true)
                .setNoticesCssStyle(cssFormat)
                .build();

        dialog.show();*/
    }
}

File: app/src/main/java/com/gigabytedevelopersinc/app/cometOTP/Activities/AboutActivity.kt
@file:Suppress("PackageName")
package com.gigabytedevelopersinc.app.cometOTP.Activities

import android.annotation.SuppressLint
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import android.view.ViewStub
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.widget.Toolbar
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import com.gigabytedevelopersinc.app.cometOTP.R
import com.gigabytedevelopersinc.app.cometOTP.Utilities.Tools
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.snackbar.Snackbar
import saschpe.android.customtabs.CustomTabsHelper
import saschpe.android.customtabs.WebViewFallback


class AboutActivity : BaseActivity() {
    private var mBottomSheetDialog: BottomSheetDialog? = null

    @SuppressLint("IntentReset")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTitle(R.string.about_activity_title)
        setContentView(R.layout.activity_container)
        val toolbar = findViewById<Toolbar>(R.id.container_toolbar)
        setSupportActionBar(toolbar)
        val stub = findViewById<ViewStub>(R.id.container_stub)
        stub.layoutResource = R.layout.content_about
        val v = stub.inflate()
        val builder = CustomTabsIntent.Builder()
        builder.setShowTitle(true)
        val params = CustomTabColorSchemeParams.Builder()
            //.setNavigationBarColor(ContextCompat.getColor(this, R.color.background))
            .setToolbarColor(ContextCompat.getColor(this, R.color.colorPrimary))
            //.setSecondaryToolbarColor(ContextCompat.getColor(activity, R.color.background))
            .build()
        builder.setColorSchemeParams(CustomTabsIntent.COLOR_SCHEME_DARK, params)
        val customTabsIntent = builder.build()
        val customTabsCompanion = CustomTabsHelper.Companion
        customTabsCompanion.addKeepAliveExtra(this, customTabsIntent.intent)
        var versionName = ""
        try {
            val packageInfo = packageManager.getPackageInfo(packageName, 0)
            versionName = packageInfo.versionName.toString()
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        val versionLayout = v.findViewById<LinearLayout>(R.id.about_layout_version)
        versionLayout.setOnClickListener {
            val thisTap = System.currentTimeMillis()
            if (thisTap - lastTap < 500) {
                taps = taps + 1
                if (currentToast != null && taps <= 7) currentToast!!.cancel()
                if (taps >= 3 && taps <= 7) currentToast = Toast.makeText(
                    baseContext, taps.toString(), Toast.LENGTH_SHORT
                )
                if (taps == 7) {
                    if (settings.specialFeatures) Snackbar.make(
                        findViewById(R.id.about),
                        R.string.about_toast_special_features_enabled,
                        Snackbar.LENGTH_LONG
                    ).show() else enableSpecialFeatures()
                }
                if (currentToast != null) currentToast!!.show()
            } else {
                taps = 0
            }
            lastTap = thisTap
        }
        val version = v.findViewById<TextView>(R.id.about_text_version)
        version.text = versionName
        val license = v.findViewById<LinearLayout>(R.id.about_layout_license)
        val changelog = v.findViewById<LinearLayout>(R.id.about_layout_changelog)
        val source = v.findViewById<LinearLayout>(R.id.about_layout_source)
        val licenses = v.findViewById<LinearLayout>(R.id.about_layout_licenses)
        license.setOnClickListener { openURI(MIT_URI) }
        changelog.setOnClickListener {
            customTabsCompanion.openCustomTab(
                this, customTabsIntent,
                changeLogUrl.toUri(),
                WebViewFallback()
            )
        }
        source.setOnClickListener { openURI(WHATSAPP_URI) }
        licenses.setOnClickListener { showLicenses() }
        val author1GitHub = v.findViewById<TextView>(R.id.about_author1_github)
        val author1Paypal = v.findViewById<TextView>(R.id.about_author1_paypal)
        author1GitHub.setOnClickListener {
            try {
                openURI(AUTHOR1_GITHUB)
            } catch (ignored: Exception) {
                copyToClipboard(AUTHOR1_GITHUB)
            }
        }
        author1Paypal.setOnClickListener {
            try {
                openURI(AUTHOR1_PAYPAL)
            } catch (ignored: Exception) {
                copyToClipboard(AUTHOR1_PAYPAL)
            }
        }
        val author2GitHub = v.findViewById<TextView>(R.id.about_author2_github)
        val author2Paypal = v.findViewById<TextView>(R.id.about_author2_paypal)
        author2GitHub.setOnClickListener {
            try {
                openURI(AUTHOR2_GITHUB)
            } catch (ignored: Exception) {
                copyToClipboard(AUTHOR2_GITHUB)
            }
        }
        author2Paypal.setOnClickListener {
            try {
                openURI(AUTHOR2_PAYPAL)
            } catch (ignored: Exception) {
                copyToClipboard(AUTHOR2_PAYPAL)
            }
        }
        val bugReport = v.findViewById<LinearLayout>(R.id.about_layout_bugs)
        bugReport.setOnClickListener {
            val feedback = Intent(Intent.ACTION_SENDTO)
            feedback.type = "text/html"
            feedback.data = "mailto:".toUri()
            feedback.putExtra(
                Intent.EXTRA_EMAIL,
                arrayOf(getString(R.string.feedback_email_address))
            )
            feedback.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.feedback_email_subject))
            feedback.putExtra(Intent.EXTRA_TEXT, getString(R.string.feedback_email_message))
            startActivity(Intent.createChooser(feedback, getString(R.string.feedback_email_title)))
        }
        val expandButton = v.findViewById<ImageView>(R.id.thumb_expand_button)
        val expand = v.findViewById<LinearLayout>(R.id.thumb_expand)
        val disclaimer = v.findViewById<TextView>(R.id.thumb_disclaimer)
        val privacyPolicy = v.findViewById<LinearLayout>(R.id.privacy_policy)

        privacyPolicy.setOnClickListener {
            val privacyPolicyIntent = Intent(this@AboutActivity, PrivacyPolicyActivity::class.java)
            resultLauncher.launch(privacyPolicyIntent)
        }
        expand.setOnClickListener {
            val open = disclaimer.visibility != View.VISIBLE
            disclaimer.visibility = if (open) View.VISIBLE else View.GONE
            expandButton.setImageResource(
                if (open) R.drawable.ic_expand_less else R.drawable.ic_expand_more
            )
        }

        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)
    }

    private val resultLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == RESULT_OK) {
            // There are no request codes
            val data: Intent? = result.data
            //doSomeOperations()
        }
    }

    @SuppressLint("InflateParams")
    private fun enableSpecialFeatures() {
        val bottomSheetLayout = layoutInflater.inflate(R.layout.bottom_sheet_special_features, null)
        bottomSheetLayout.findViewById<View>(R.id.button_no)
            .setOnClickListener { mBottomSheetDialog!!.dismiss() }
        bottomSheetLayout.findViewById<View>(R.id.button_yes).setOnClickListener {
            mBottomSheetDialog!!.dismiss()
            settings.specialFeatures = true
            Snackbar.make(
                findViewById(R.id.about), R.string.about_toast_special_features,
                Snackbar.LENGTH_LONG
            ).show()
        }
        mBottomSheetDialog = BottomSheetDialog(this)
        mBottomSheetDialog!!.setContentView(bottomSheetLayout)
        mBottomSheetDialog!!.setCancelable(false)
        mBottomSheetDialog!!.show()
    }

    // Go back to the main activity
    /*override fun onSupportNavigateUp(): Boolean {
        finish()
        return true

    }*/

    private val onBackPressedCallback = object: OnBackPressedCallback(true) {
        override fun handleOnBackPressed() {
            finish()
        }
    }

    private fun openURI(uri: String?) {
        val openURI = Intent(Intent.ACTION_VIEW)
        if (uri != null) {
            openURI.data = uri.toUri()
        }
        startActivity(openURI)
    }

    private fun copyToClipboard(uri: String?) {
        val clipboard = getSystemService(CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("CometOTP", uri)
        clipboard.setPrimaryClip(clip)
        Snackbar.make(
            findViewById(R.id.about),
            getString(R.string.about_toast_copied_to_clipboard),
            Snackbar.LENGTH_LONG
        ).show()
        //Toast.makeText(this, getString(R.string.about_toast_copied_to_clipboard), Toast.LENGTH_SHORT).show();
    }

    private fun showLicenses() {
        val licensesIntent = Intent(this, LicensesActivity::class.java)
        resultLauncher.launch(licensesIntent)
        /*String backgroundColor = Tools.getCSSRGBAString(Tools.getThemeColor(this, R.attr.colorBackgroundFloating));
        String textColor = Tools.getCSSRGBAString(Tools.getThemeColor(this, android.R.attr.textColorPrimary));
        String textColorSecondary = Tools.getCSSRGBAString(Tools.getThemeColor(this, android.R.attr.textColorSecondary));
        String cssFormat = getString(R.string.custom_notices_style, backgroundColor, textColor, textColorSecondary);
        LicensesDialog dialog = new LicensesDialog.Builder(this)
                .setNotices(R.raw.licenses)
                .setTitle(R.string.about_label_licenses)
                .setShowFullLicenseText(false)
                .setIncludeOwnLicense(true)
                .setNoticesCssStyle(cssFormat)
                .build();

        dialog.show();*/
    }

    companion object {
        private const val GITHUB_URI = "https://github.com/gigabytedevelopers/CometOTP"
        private const val WHATSAPP_URI = "https://chat.whatsapp.com/HvUjQXAkMne3hjSiy1sUpc"

        // private static final String CHANGELOG_URI = GITHUB_URI + "/blob/master/CHANGELOG.md";
        private const val MIT_URI = "$GITHUB_URI/blob/master/LICENSE.txt"
        private const val AUTHOR1_GITHUB = "https://github.com/gigabytedevelopers"
        private const val AUTHOR1_PAYPAL = "https://paypal.me/gigabtedevelopers"

        private const val AUTHOR2_GITHUB = "https://github.com/enwokoma"
        private const val AUTHOR2_PAYPAL = "https://www.paypal.com/donate/?hosted_button_id=G35ZM2EKPVZ9Y"
        private const val changeLogUrl = "https://gigabytedevelopers.com/apps/cometotp/changelog"

        // private static final String BUGREPORT_URI = GITHUB_URI + "/issues";
        val imageResources = intArrayOf(
            R.id.aboutImgVersion, R.id.aboutImgLicense, R.id.aboutImgChangelog, R.id.aboutImgSource,
            R.id.aboutImgOpenSource, R.id.aboutImgAuthor1, R.id.aboutImgAuthor2, R.id.aboutImgBugs
        )
        var lastTap: Long = 0
        var taps = 0
        var currentToast: Toast? = null
    }
}